dgisim/tests/test_dices.py

```diff
@@ -167,101 +168,438 @@ def test_just_satisfy_typical_all_omni(self):
+        dices = ActualDices({
+            Element.HYDRO: 1,
+            Element.ELECTRO: 2,
+            Element.ANEMO: 1,
+            Element.GEO: 1,
+            Element.CRYO: 2,
+            Element.OMNI: 1,
+            Element.DENDRO: 2,
+            Element.PYRO: 1,
+        })
+        keys = tuple(dices.readonly_dices_ordered(None).keys())
+        expected_order = (
+            Element.OMNI,
+            Element.CRYO,
+            Element.ELECTRO,
+            Element.DENDRO,
+            Element.HYDRO,
+            Element.PYRO,
+            Element.GEO,
+            Element.ANEMO,
+        )
+        self.assertEqual(
+            keys,
+            expected_order,
+        )
+
+    def test_ordered_actual_dices_with_characters(self):
+        dices = ActualDices({
+            Element.HYDRO: 1,
+            Element.ELECTRO: 2,
+            Element.ANEMO: 1,
+            Element.GEO: 1,
+            Element.CRYO: 2,
+            Element.OMNI: 1,
+            Element.DENDRO: 2,
+            Element.PYRO: 1,
+        })
+        from dgisim.src.card.cards import Cards
+        from dgisim.src.character.character import AratakiItto, Klee, Keqing
+        from dgisim.src.character.characters import Characters
+        from dgisim.src.mode import DefaultMode
+        from dgisim.src.state.player_state import PlayerState
+        player_state = PlayerState.from_chars_cards(
+            DefaultMode(),
+            Characters.from_iterable(
+                [AratakiItto, Klee, Keqing]
+            ).factory().active_character_id(2).build(),
+            Cards({}),
+        )
+        keys = tuple(dices.readonly_dices_ordered(player_state).keys())
+        expected_order = (
+            Element.OMNI,
+            Element.ELECTRO,
+            Element.PYRO,
+            Element.GEO,
+            Element.CRYO,
+            Element.DENDRO,
+            Element.HYDRO,
+            Element.ANEMO,
+        )
+        self.assertEqual(
+            keys,
+            expected_order,
+        )
+
     # test_smart_selection template
     # def test_smart_selection(self):
     #     input_actual_dices: ActualDices = ActualDices()
     #     input_abstract_dices: AbstractDices = AbstractDices()
-    #     input_precedence: List[Set[Element]] = []
-    #     #
-    #     # self.assertEqual()
+    #     input_precedence: List[set[Element]] = []
+    #
+    #     self.assertEqual()
+    #
+    # def test_smart_selection_TEMPLATE(self, k: int, n: int):
 
-    # def test_smart_selection_omni(self, k: int, n: int):
     def test_smart_selection_omni(
-            self, k: int = 3, n: int = 2, element: Element = Element.GEO):
-        input_actual_dices: ActualDices = ActualDices(dices={element: k})
-        input_abstract_dices: AbstractDices = AbstractDices(
-            dices={Element.OMNI: n})
-        input_precedence: List[Set[Element]] = []
+            self, element: Element = Element.GEO):
+        """test if element can fill OMNI"""
```

Comment:
I normally comment with preceding and trailing spaces like:
```py
""" my comments or documentation """
```

Reply:
done